In [162]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in c:\users\seble\anaconda3\lib\site-packages (0.0)


In [163]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [164]:
import pandas as pd
%matplotlib inline
from IPython.display import Image, SVG
import matplotlib.pyplot as plt
import numpy as np
np.random.seed(0)


In [165]:
# Filepaths, numpy, and Tensorflow
import os
import numpy as np
import tensorflow as tf

In [166]:
# Sklearn scaling
from sklearn.preprocessing import MinMaxScaler

In [167]:
# Keras
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Dense
from tensorflow.keras.datasets import mnist

# Read the CSV and Perform Basic Data Cleaning

In [168]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna()
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [169]:
df. shape

(6991, 41)

# Select your features (columns)

In [170]:
# Set features. This will also be used as your x values.
selected_features = df[['koi_period', 'koi_time0bk', 'koi_slogg', 'koi_srad', 'koi_kepmag']]

In [171]:
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [172]:
df.tail()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  \
6986  FALSE POSITIVE              0              0              0   
6987  FALSE POSITIVE              0              1              1   
6988       CANDIDATE              0              0              0   
6989  FALSE POSITIVE              0              0              1   
6990  FALSE POSITIVE              0              0              1   

      koi_fpflag_ec  koi_period  koi_period_err1  koi_period_err2  \
6986              1    8.589871     1.846000e-04    -1.846000e-04   
6987              0    0.527699     1.160000e-07    -1.160000e-07   
6988              0    1.739849     1.780000e-05    -1.780000e-05   
6989              0    0.681402     2.430000e-06    -2.430000e-06   
6990              1    4.856035     6.360000e-05    -6.360000e-05   

      koi_time0bk  koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  \
6986   132.016100           0.01570  ...            -152      4.296   
6987   131.705093           0.00017  ...            -166      4.529   
6988   133.001270           0.00769  ...            -220      4.444   
6989   132.181750           0.00285  ...            -236      4.447   
6990   135.993300           0.01080  ...            -225      4.385   

      koi_slogg_err1  koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2  \
6986           0.231          -0.189     1.088          0.313         -0.228   
6987           0.035          -0.196     0.903          0.237         -0.079   
6988           0.056          -0.224     1.031          0.341         -0.114   
6989           0.056          -0.224     1.041          0.341         -0.114   
6990           0.054          -0.216     1.193          0.410         -0.137   

             ra        dec  koi_kepmag  
6986  298.74921  46.973351      14.478  
6987  297.18875  47.093819      14.082  
6988  286.50937  47.163219      14.757  
6989  294.16489  47.176281      15.385  
6990  297.00977  47.121021      14.826  

[5 rows x 41 columns]

# Create a Train Test Split

Use `koi_disposition` for the y values

In [173]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

#assign x and y values
X = df.drop("koi_disposition", axis=1)
y = df["koi_disposition"]

print(X.shape, y.shape)

(6991, 40) (6991,)


In [174]:
#split training and testing data
from sklearn.model_selection import train_test_split

X = pd.get_dummies(X)

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=84, stratify =y)


In [175]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
3309              0              0              0              0   10.471619   
5173              0              1              0              0    4.206070   
6070              1              0              0              0   37.537859   
2421              0              0              0              0  105.881767   
6239              1              0              1              1   25.952291   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
3309         0.000153        -0.000153   136.287700           0.01180   
5173         0.000003        -0.000003   132.248423           0.00050   
6070         0.001282        -0.001282   159.736400           0.02760   
2421         0.000131        -0.000131   156.979350           0.00101   
6239         0.000846        -0.000846   154.868900           0.02050   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
3309          -0.01180  ...             -78      4.083           0.182   
5173          -0.00050  ...            -343      3.736           0.432   
6070          -0.02760  ...            -253      4.429           0.065   
2421          -0.00101  ...             -79      3.757           0.078   
6239          -0.02050  ...            -197      4.547           0.032   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
3309          -0.084     1.574          0.240         -0.331  282.09314   
5173          -0.108     3.185          0.789         -1.465  298.47604   
6070          -0.195     1.082          0.331         -0.142  289.21732   
2421          -0.052     2.499          0.233         -0.311  291.49722   
6239          -0.179     0.866          0.217         -0.078  293.70291   

            dec  koi_kepmag  
3309  42.160271      14.486  
5173  41.079910      11.684  
6070  48.204311      15.769  
2421  42.728481      10.775  
6239  38.956291      15.999  

[5 rows x 40 columns]

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [176]:
# Scale your data
from sklearn.preprocessing import MinMaxScaler


In [177]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=1, stratify=y)

In [178]:
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Step 1: Label-encode data set
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)

# Step 2: Convert encoded labels to one-hot-encoding
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

In [179]:
# Create an empty sequential model
model1.score = Sequential()
#model1 = Sequential()
#model1.add(Dense(100, activation='relu', input_dim=X_train.shape[1]))
#model1.add(Dense(100, activation='relu'))
#model1.add(Dense(num_classes, activation='softmax'))

In [180]:
print(f"Training Data Score: {model1.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model1.score(X_test_scaled, y_test)}")


If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

Training Data Score: [[0.         0.         1.         ... 0.60380167 0.5514255  0.73004717]
 [0.         1.         0.         ... 0.47678995 0.5652877  0.71780664]
 [0.         1.         1.         ... 0.9689742  0.5368881  0.58431894]
 ...
 [0.         1.         0.         ... 0.47624844 0.22670393 0.27433628]
 [0.         0.         1.         ... 0.56817526 0.10331628 0.68133324]
 [1.         0.         0.         ... 0.47496465 0.07118876 0.67703253]]
Testing Data Score: [[0.         0.        

In [181]:
model1

In [182]:
y_test.head()

1981         CANDIDATE
5609    FALSE POSITIVE
532     FALSE POSITIVE
6558         CANDIDATE
1249    FALSE POSITIVE
Name: koi_disposition, dtype: object

# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [193]:
# Create the SVC Model
from sklearn.svm import SVC 
model2 = SVC(kernel='linear')
model2


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [194]:
# Create the GridSearchCV model
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1, 5, 10],
              'gamma': [0.0001, 0.001, 0.01]}
grid = GridSearchCV(model2, param_grid, verbose=3)

In [195]:
# Train the model with GridSearch
grid.fit(X_train_scaled, y_train)

C:\Users\seble\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.853, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.2s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.837, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.5s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.828, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.853, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.837, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.828, total=   0.3s
[CV] C=1, gamma=0.01 .................................................
[CV] ..................... C=1, gamma=0.01, score=0.853, total=   0.3s
[CV] C=1, gamma=0.01 .................................................
[CV] ..................... C=1, gamma=0.01, score=0.837, total=   0.3s
[CV] C=1, gamma=0.01 .................................................
[CV] ..................... C=1, gamma=0.01, score=0.828, total=   0.3s
[CV] C=5, gamma=0.0001 ...............................................
[CV] .

[Parallel(n_jobs=1)]: Done  27 out of  27 | elapsed:    8.9s finished


GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='linear',
                           max_iter=-1, probability=False, random_state=None,
                           shrinking=True, tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid={'C': [1, 5, 10], 'gamma': [0.0001, 0.001, 0.01]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [196]:
# List the best parameters for this dataset
print(grid.best_params_)

{'C': 10, 'gamma': 0.0001}


In [197]:
# List the best score
print(grid.best_score_)

0.8661071905397673


In [198]:
# List the best score
print(grid.best_score_)

0.8661071905397673


In [199]:
# Make predictions with the hypertuned model
#predictions = grid.predict(X_test)
predictions_GridSearch = grid.predict(X_test_scaled)



In [200]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions_GridSearch,
                            target_names=['CANDIDATE', 'CONFIRMED', 'FALSE POSITIVE']))

                precision    recall  f1-score   support

     CANDIDATE       0.85      0.64      0.73       422
     CONFIRMED       0.72      0.86      0.79       450
FALSE POSITIVE       0.98      1.00      0.99       876

      accuracy                           0.88      1748
     macro avg       0.85      0.83      0.83      1748
  weighted avg       0.88      0.88      0.87      1748



# Save the Model

In [201]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'Seble_model1.sav'




In [205]:
#joblib.dump(model1, 'Seble_model1.sav')

# conclusion:   Hyperparameter Tuning model shows accuracey ranging between 83-88%. The Hypertuning the SVM model achieves significantly higher accuracy of 88%, the model seems fiting for the sample data.  